src/nvim/syntax.c

```diff
@@ -6944,6 +6945,9 @@ void do_highlight(const char *line, const bool forceit, const bool init)
         // redraw below will still handle usages of guibg=fg etc.
         ui_default_colors_set();
       }
+      item = &HL_TABLE()[idx];
```

Comment:
the vim patch has a comment which is a useful hint:

    item = &HL_TABLE()[idx];  // table may have changed